---
title: "cgmisc: package tutorial"
author: "Marcin Kierczak, Jagoda Jablonska, Simon Forsberg, Matteo Bianchi, Katarina Tengvall, Mats Pettersson, Jennifer Meadows, Patric Jern, Orjan Carlborg, Kerstin Lindblad-Toh"
date: "18 Feb 2015"
output: word_document
---

```{r setup, echo=F, results='hide', message=F, warning=TRUE, eval=TRUE}
opar <- par()
par(cex=.5, cex.axis=.7, cex.lab=.7, cex.main=.7, cex.sub=.7, las=1, bty='n')
```

**cgmisc** is an R package for enhanced genome-wide association studies (GWAS) and visualisation. This document aims at guiding you through the installation process and to demonstrate package capabilities in a series of practical examples based on an example data included in the package.

# Package installation
The **cgmisc** package can be installed in the same way as any other R package. One way is to issue the following command in R console:

```{r installing, echo=TRUE, results='show', message=TRUE, warning=FALSE, eval=F}
install.packages("cgmisc")
```
Other possibilities include using graphical user interface (GUI) of, e.g. R console or RStudio.

After the package has been installed, to use the package, it is necessary to load it into environment:
```{r, echo=TRUE, results='show', message=TRUE, warning=FALSE, eval=T}
library("cgmisc")
```

# Loading data
Whenever possible, the **cgmisc** package uses data structures used by the GenABEL (Aulchenko et al., 2007) package. In particular, the ```gwaa.data-class``` and the ```gwaa.scan-class``` structures are used.
The package is shipped with an example dataset called **cgmisc_data** that contains genotyping data (Illumina, canFam2) for N=207 German shepherds originally collected for the project described in (Tengvall et al., 2012). The phenotypes, though, have been simulated in order to be able to illustrate various features of **cgmisc**. To load the example dataset, use the following command:

```{r load_data, echo=TRUE, results='show', message=TRUE, warning=FALSE, eval=T}
data("data")
```

# Example analyses
In order to illustrate how to use particular functions, we will perform a very much simplified GWAS analysis. We begin by initial quality control where we prune the data with per marker of per individual call rates below 95%. Based on 2000 randomly selected markers, we remove one (with lower call rate) from each pair of too similar (more than 95% similarity) individuals. We also set very low ($10^{-3}$) threshold for pruning on minor allele frequency (in practise only the monomorphic markers will be removed) and turn off checks based on the departure form Hardy-Weinberg equilibrium (p.level=10e-18)
```{r qc1, echo=TRUE, results='show', message=TRUE, warning=FALSE, eval=T}
qc1 <- check.marker(data, callrate = .95, perid.call = .95, ibs.threshold = .95, ibs.mrk=2000, ibs.exclude="lower", p.level=10e-18, maf=1e-3)
data.qc1 <- data[qc1$idok, qc1$snpok]
```

Next, we analyse population structure by means of genomic-kinship:
```{r gkin, echo=TRUE, results='show', message=TRUE, warning=FALSE, eval=T}
autosomal <- which(data.qc1@gtdata@chromosome != 39)
data.qc1.gkin <- ibs(data.qc1, snpsubset = autosomal, weight = 'freq')
data.qc1.dist <- as.dist(0.5 - data.qc1.gkin)
data.qc1.mds <- cmdscale(data.qc1.dist)
plot(data.qc1.mds, pch=19, xlab="MDS1", ylab="MDS2")
```

We can see that there is possible population structure here. We should investigate this further, but for our purposes, let's just run simple K-means clustering with the number of clusters *a priori* set to $K = 2$
```{r kclust, echo=TRUE, results='show', message=TRUE, warning=FALSE, eval=T}
kclust <- kmeans(data.qc1.mds, centers = 2)
plot(data.qc1.mds, pch=19, xlab="MDS1", ylab="MDS2", col=kclust$cluster)
pop <- kclust$cluster
```

Having defined subpopulations, we can proceed to association analyses using mixed model with genomic kinship as random effect.
```{r h2h, echo=TRUE, results='show', message=TRUE, warning=FALSE, eval=T}
h2h <- polygenic_hglm(formula = ct ~ sex, data.qc1.gkin, data.qc1)
mm <- mmscore(h2h, data.qc1, strata = pop)
plot(mm, cex=.5, pch=19, col=c("darkgrey","grey"))
```

As we can see, there is a very strong association signal on chromosome 2. We can examine it a bit closer using the ```plot.manhattan.ld``` function.

# Visualization and analyses of linkage structure

Say, we would like to zoom in on chromosome 2 and visualise LD to the top-associated marker. First, we need the name and cooridinates of the marker:
```{r summary.top, echo=TRUE, results='show', message=TRUE, warning=FALSE, eval=T}
summary(mm, top=1)
```
We see that the top-associated marker is **BICF2S2365880** and its position is **38256927bp**. We will zoom in on a 2Mbp region centered on the marker:
```{r plot.manhattan.ld, echo=TRUE, results='show', message=TRUE, warning=FALSE, eval=T}
plot.manhattan.LD(data = data.qc1, gwas.result = mm, chr = 2, region = c(37256927,39256927), index.snp = "BICF2S2365880", legend.pos=c(38.9e6, 15), bonferroni = F, mafThreshold = 1e-3)
```
It is also possible to identify regions of interest using clumping procedure as described in PLINK documentation [cite]:
```{r clumping, echo=TRUE, results='show', message=TRUE, warning=FALSE, eval=T}
clumps <- clump.markers(data.qc1, mm, 2)
plot.clumps(mm, clumps, chr = 2, c(37256927,39256927))
```

To visualise LD decay on chromosome 2, one can call the ```plot.ld.decay``` function.
```{r LD.decay, echo=TRUE, results='show', message=TRUE, warning=FALSE, eval=T}
#plot.LD.decay(data.qc1[,data.qc1@gtdata@chromosome==2], N=100, dmin = 1e2, dmax = 1e3)
```

